tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift

```diff
@@ -233,38 +255,299 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+          .withCString(encodedAs: UTF16.self) {
+            // Check that the dll file exists
+            var faAttributes: WIN32_FILE_ATTRIBUTE_DATA = .init()
+            guard GetFileAttributesExW($0, GetFileExInfoStandard, &faAttributes),
+              faAttributes.dwFileAttributes & DWORD(FILE_ATTRIBUTE_REPARSE_POINT) == 0
+            else {
+              print("\($0) doesn't exist")
+              return nil
+            }
+            // Allocate memory in the remote process
+            let szLength = SIZE_T(wcslen($0) * MemoryLayout<WCHAR>.size + 1)
+            guard
+              let allocation = VirtualAllocEx(
+                self.process, nil, szLength,
+                DWORD(MEM_COMMIT), DWORD(PAGE_READWRITE))
+            else {
+              print("VirtualAllocEx failed \(GetLastError())")
+              return nil
+            }
+            // Write the path in the allocated memory
+            if !WriteProcessMemory(self.process, allocation, $0, szLength, nil) {
+              print("WriteProcessMemory failed \(GetLastError())")
+              VirtualFreeEx(self.process, allocation, 0, DWORD(MEM_RELEASE))
+              return nil
+            }
+
+            return allocation
+          }
+      }
+  }
+
+  private func unloadDllAndPathRemote(
+    dwProcessId: DWORD, hKernel32: HMODULE!,
+    dllPathRemote: UnsafeMutableRawPointer
+  ) -> Bool {
+    // Get the dll module handle in the remote process to use it to
+    // unload it below.
+    // GetExitCodeThread returns a DWORD (32-bit) but the HMODULE
+    // returned from LoadLibraryW is a 64-bit pointer and may be truncated.
+    // So, search for it using the snapshot instead.
+    var hDllModule = HMODULE(bitPattern: 0)
+    let hModuleSnapshot: HANDLE =
+      CreateToolhelp32Snapshot(DWORD(TH32CS_SNAPMODULE), dwProcessId)
+    if hModuleSnapshot == INVALID_HANDLE_VALUE {
+      print("CreateToolhelp32Snapshot failed \(GetLastError())")
+      return false
+    }
+    defer { CloseHandle(hModuleSnapshot) }
+    var entry: MODULEENTRY32W = MODULEENTRY32W()
+    entry.dwSize = DWORD(MemoryLayout<MODULEENTRY32W>.size)
+    guard Module32FirstW(hModuleSnapshot, &entry) else {
+      print("Module32FirstW failed \(GetLastError())")
+      return false
+    }
+    repeat {
+      let module: String = withUnsafePointer(to: entry.szModule) {
+        $0.withMemoryRebound(
+          to: WCHAR.self,
+          capacity: MemoryLayout.size(ofValue: $0) / MemoryLayout<WCHAR>.size
+        ) {
+          String(decodingCString: $0, as: UTF16.self)
+        }
       }
+      if module == "SwiftInspectClient.dll" {
+        hDllModule = entry.hModule
+      }
+    } while Module32NextW(hModuleSnapshot, &entry)
+    if hDllModule == HMODULE(bitPattern: 0) {
+      print("Failed to find the dll")
+      return false
+    }
+
+    // Unload the dll from the remote process
+    let unloadHookAddr = unsafeBitCast(
```

Comment:
Is this correct?  This is using the address of `Kernel32.FreeLibrary` in the `swift-inspect` process but invoking it in the remote process, which is not guaranteed to be loaded at the same address right?  I don't think that this is safe to do.  We need to have the function be bound in the process.  The simplest way to do this is to have a secondary entry point that we can invoke in the injected module that can unload itself.

Reply:
I think that kernel32 is special and guaranteed to be loaded at the same address across processes.

https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-entry-point-function

> Because Kernel32.dll is guaranteed to be loaded in the process address space when the entry-point function is called...

Reply:
I'm not sure I read it the same way - it is that it is loaded before the entry point executes, it doesn't guarantee that the module is loaded at the same address, ASLR can relocate the base address.

Reply:
Yeah, thatwas a misquote, sorry.

I'm reading that kernel32 is loaded at the same address across processes, however. For example, http://www.nynaeve.net/?p=198

Isn't that assumed and necessary when we call `LoadLibraryW` earlier?

Reply:
Yes, that is a mistake as well.  Discussed this offline, and we will use the toolhelp32 to snapshot and find the `HMODULE` associated with kernel32.dll in the image and use `GetProcAddress` to get the proper location.